File: swift/Sources/HermesAPILiveScenarios/LiveScenarios.swift
import Foundation
import HermesAPI

/// Live scenario inputs, read from `HERMES_LIVE_*` variables that the harness sets.
public struct LiveScenarioEnvironment: Sendable {
    public let url: URL
    public let auth: any HermesAuth
    public let token: String?
    public let lifecycle: Bool
    public let control: URL?

    public init(_ values: [String: String]) throws {
        guard let rawURL = values["HERMES_LIVE_URL"], let url = URL(string: rawURL) else {
            throw LiveScenarioError("HERMES_LIVE_URL is required")
        }
        self.url = url
        if let ticket = values["HERMES_LIVE_TICKET"], !ticket.isEmpty {
            auth = StaticTicketAuth(ticket: ticket)
            token = nil
        } else if let token = values["HERMES_LIVE_TOKEN"], !token.isEmpty {
            auth = LocalTokenAuth(token: token)
            self.token = token
        } else {
            throw LiveScenarioError("HERMES_LIVE_TICKET or HERMES_LIVE_TOKEN is required")
        }
        lifecycle = values["HERMES_LIVE_LIFECYCLE"] == "1"
        control = values["HERMES_LIVE_CONTROL"].flatMap(URL.init(string:))
    }
}

public struct LiveScenarioError: Error, CustomStringConvertible, LocalizedError {
    public let description: String

    public init(_ description: String) { self.description = description }

    public var errorDescription: String? { description }
}

/// Fixture replies produced by the harness model stub (`harness/stub_llm.py`).
enum Fixture {
    static let reply = "HermesAPI fixture reply."
    static let clarifyPrompt = "Ask which release channel to use for HermesAPI."
    static let clarifyReply = "HermesAPI stable release selected."
    static let approvalPrompt = "Try the fixture cleanup command and report whether it was approved."
    static let approvalCommand = "rm -rf /tmp/hermes-api-fixture-approval-target"
    static let approvalReply = "HermesAPI approval denied as expected."
    static let reconnectPrompt = "Stream the HermesAPI reconnect fixture slowly."
    static let reconnectReply = (1...16).map { String(format: "part%02d", $0) }.joined(separator: " ")
}

public enum LiveScenarios {
    /// Runs the smoke scenario and, when the harness exposes its control endpoint, the reconnect scenarios.
    public static func run(_ environment: LiveScenarioEnvironment) async throws {
        let observations = LiveObservations()
        try await smoke(environment, observations: observations)
        guard let control = environment.control else { return }
        let faults = FaultControl(base: control)
        if environment.lifecycle {
            try await reconnect(environment, faults: faults, observations: observations)
        }
        // Only a fully passing run reports what it exercised.
        try await faults.report(observations.report())
    }

    static func smoke(_ environment: LiveScenarioEnvironment, observations: LiveObservations) async throws {
        let gateway = HermesGateway(configuration: .init(
            baseURL: environment.url, auth: environment.auth,
            transport: ObservingTransport(inner: URLSessionGatewayTransport(), observations: observations)))
        await gateway.setServerRequestHandler { request in
            switch request {
            case .clarify(let params):
                guard case .value(let questions) = params.questions,
                      questions.count == 1,
                      questions[0].question == "Which release channel?" else {
                    throw LiveScenarioError("Unexpected clarification request")
                }
                return .clarify(ClarifyResult(answers: [questions[0].qid: "Stable"]))
            case .approval(let params):
                guard params.command == Fixture.approvalCommand else {
                    throw LiveScenarioError("Unexpected approval command")
                }
                return .approval(ApprovalResult(choice: .deny))
            default:
                throw LiveScenarioError("Unexpected server request")
            }
        }
        do {
            try await gateway.connect()
            let result = try await gateway.ping(PingParams())
            guard result.pong else { throw LiveScenarioError("Gateway ping returned false") }
            _ = try await gateway.gateway.capabilities(PingParams())
            if environment.lifecycle {
                try await lifecycle(gateway, environment: environment)
            }
            await gateway.disconnect()
        } catch {
            await gateway.disconnect()
            throw error
        }
    }

    private static func lifecycle(_ gateway: HermesGateway, environment: LiveScenarioEnvironment) async throws {
        // Hermes broadcasts sessions.changed (at most every 2 s) once a turn writes the session store.
        let listChanges = gateway.events()
        let session = try await gateway.session.create(.init(
            cwd: .value("/tmp"), title: .value("HermesAPI live session"), closeOnDisconnect: true))
        let greeting = try await runTurn(gateway, sessionID: session.sessionId,
                                         prompt: "Reply with a short greeting.", tool: nil)
        try greeting.expect(reply: Fixture.reply)
        try await withDeadline(.seconds(15), "a sessions.changed broadcast") {
            for await event in listChanges where event.type == "sessions.changed" {
                guard case .sessionsChanged = event.payload else { throw LiveScenarioError("Untyped sessions.changed") }
                return
            }
            throw LiveScenarioError("Event stream ended before sessions.changed")
        }
        let clarified = try await runTurn(gateway, sessionID: session.sessionId,
                                          prompt: Fixture.clarifyPrompt, tool: "clarify")
        try clarified.expect(reply: Fixture.clarifyReply)
        let denied = try await runTurn(gateway, sessionID: session.sessionId,
                                       prompt: Fixture.approvalPrompt, tool: "terminal")
        try denied.expect(reply: Fixture.approvalReply)
        guard case .object(let result)? = denied.toolResult,
              result["status"] == .string("blocked"),
              result["exit_code"] == .integer(-1) else {
            throw LiveScenarioError("Denied terminal command was not blocked")
        }
        _ = try await gateway.session.list(.init())
        let closed = try await gateway.session.close(.init(sessionId: session.sessionId))
        guard closed.closed else { throw LiveScenarioError("Gateway session did not close") }
        guard let token = environment.token else { throw LiveScenarioError("REST smoke needs the local token") }
        let rest = HermesREST(configuration: .init(
            baseURL: environment.url, headers: { ["X-Hermes-Session-Token": token] }))
        let voice = try await rest.audio.voiceLiveStatus(profile: "default")
        guard voice.ok, voice.mode == "chained", !voice.model.isEmpty, !voice.voice.isEmpty else {
            throw LiveScenarioError("Unexpected voice status")
        }
        let profile = try await rest.profiles.active()
        guard profile.active == "default", profile.current == "default" else {
            throw LiveScenarioError("Unexpected active profile")
        }
        let selected = try await rest.profiles.setActive(body: .init(name: "default"))
        guard selected.ok, selected.active == "default" else {
            throw LiveScenarioError("Could not select the default profile")
        }
        let count = try await rest.sessions.emptyCount(profile: "default")
        guard count.count >= 0 else { throw LiveScenarioError("Invalid empty session count") }
    }

    /// Loses the socket mid-stream, silently, and with a question open, then resumes the session after a
    /// server restart. The heartbeat is shortened so a stalled socket is detected within seconds.
    static func reconnect(
        _ environment: LiveScenarioEnvironment, faults: FaultControl, observations: LiveObservations
    ) async throws {
        let gateway = HermesGateway(configuration: .init(
            baseURL: environment.url, auth: environment.auth,
            transport: ObservingTransport(inner: URLSessionGatewayTransport(), observations: observations),
            reconnectDelay: { _ in .milliseconds(250) },
            heartbeatInterval: .seconds(1), heartbeatDeadline: .seconds(4)))
        let clarifications = Counter()
        await gateway.setServerRequestHandler { request in
            guard case .clarify(let params) = request,
                  case .value(let questions) = params.questions, questions.count == 1 else {
                throw LiveScenarioError("Unexpected server request during reconnect")
            }
            if await clarifications.increment() == 1 {
                // Drop the socket while the question is open; the rebind must re-deliver it.
                try await faults.drop(holdMilliseconds: 1_000)
                try await Task.sleep(for: .seconds(60))
            }
            return .clarify(ClarifyResult(answers: [questions[0].qid: "Stable"]))
        }
        let states = StateLog()
        let stateStream = gateway.connectionStates()
        let stateTask = Task {
            for await state in stateStream { await states.append(state) }
        }
        defer { stateTask.cancel() }
        do {
            try await gateway.connect()
            let session = try await gateway.session.create(.init(
                cwd: .value("/tmp"), title: .value("HermesAPI reconnect session"), closeOnDisconnect: false))

            let streamed = try await runTurn(
                gateway, sessionID: session.sessionId, prompt: Fixture.reconnectPrompt, tool: nil,
                onFirstDelta: { try await faults.drop(holdMilliseconds: 1_500) })
            try streamed.expect(reply: Fixture.reconnectReply)
            guard streamed.replayedEvents > 0 else {
                throw LiveScenarioError("No event reached the client through reconnect replay")
            }
            try streamed.expectContiguousSequence()
            guard await states.reconnects() >= 1 else { throw LiveScenarioError("Gateway never reconnected") }

            // A stalled socket reports nothing; the default heartbeat must replace it before the
            // server reaps the session, which the next turn on the same session then proves.
            let reconnectsBeforeStall = await states.reconnects()
            try await faults.blackhole()
            try await withDeadline(.seconds(20), "the heartbeat to replace a silently stalled socket") {
                while await states.reconnects() <= reconnectsBeforeStall {
                    try await Task.sleep(for: .milliseconds(100))
                }
            }

            let clarified = try await runTurn(gateway, sessionID: session.sessionId,
                                              prompt: Fixture.clarifyPrompt, tool: "clarify")
            try clarified.expect(reply: Fixture.clarifyReply)
            try clarified.expectContiguousSequence()
            guard await clarifications.value == 2 else {
                throw LiveScenarioError("The open clarification was not re-delivered after reconnect")
            }

            // A restart loses every live session; the gateway resumes this one from Hermes' storage.
            let reconnectsBeforeRestart = await states.reconnects()
            let recoveries = gateway.sessionRecoveries()
            try await faults.restart()
            let recovery = try await withDeadline(.seconds(60), "the restarted server's session recovery") {
                for await recovery in recoveries { return Optional(recovery) }
                return nil
            }
            guard case .resumed(let previous, let resumedID, let storedID)? = recovery,
                  previous == session.sessionId, storedID == session.storedSessionId else {
                throw LiveScenarioError("Expected \(session.sessionId) to be resumed, got \(String(describing: recovery))")
            }
            try await withDeadline(.seconds(60), "the gateway to reconnect after the restart") {
                while await !states.isConnected(afterReconnects: reconnectsBeforeRestart) {
                    try await Task.sleep(for: .milliseconds(100))
                }
            }
            let resumed = try await runTurn(gateway, sessionID: resumedID,
                                            prompt: "Reply with a short greeting.", tool: nil)
            try resumed.expect(reply: Fixture.reply)
            await gateway.disconnect()
        } catch {
            await gateway.disconnect()
            throw error
        }
    }

    private static func runTurn(
        _ gateway: HermesGateway, sessionID: String, prompt: String, tool: String?,
        onFirstDelta: (@Sendable () async throws -> Void)? = nil
    ) async throws -> Turn {
        // Subscribe before submitting so no event of the turn can precede the subscription.
        let events = gateway.events()
        let submission = try await gateway.prompt.submit(.init(sessionId: sessionID, text: .string(prompt)))
        guard submission.status != nil else { throw LiveScenarioError("Gateway rejected the prompt") }
        return try await withDeadline(.seconds(45), "the turn for \"\(prompt)\" to complete") {
            var turn = Turn(tool: tool)
            for await event in events where event.sessionID == sessionID {
                if let seq = event.seq { turn.sequence.append(seq) }
                if event.replayed { turn.replayedEvents += 1 }
                switch event.payload {
                case .messageStart:
                    turn.sawStart = true
                case .toolStart(let payload) where payload.name == tool:
                    turn.sawToolStart = true
                case .toolComplete(let payload) where payload.name == tool:
                    turn.sawToolComplete = true
                    turn.toolResult = payload.result
                case .messageDelta(let payload):
                    let first = turn.streamed.isEmpty
                    turn.streamed += payload.text
                    if first, let onFirstDelta { try await onFirstDelta() }
                case .messageComplete(let payload):
                    if case .string(let text)? = payload.text { turn.reply = text }
                    return turn
                case .error:
                    throw LiveScenarioError("Gateway reported an error during \"\(prompt)\"")
                default:
                    continue
                }
            }
            throw LiveScenarioError("Event stream ended during \"\(prompt)\"")
        }
    }
}

private struct Turn: Sendable {
    let tool: String?
    var sawStart = false
    var sawToolStart = false
    var sawToolComplete = false
    var toolResult: JSONValue?
    var streamed = ""
    var reply: String?
    var sequence: [Int] = []
    var replayedEvents = 0

    init(tool: String?) { self.tool = tool }

    func expect(reply expected: String) throws {
        guard let reply, reply == expected else {
            throw LiveScenarioError("Expected reply \"\(expected)\", got \"\(reply ?? "nothing")\"")
        }
        guard sawStart else { throw LiveScenarioError("No message.start before \"\(expected)\"") }
        guard streamed.trimmingCharacters(in: .whitespacesAndNewlines) == reply else {
            throw LiveScenarioError("Streamed text \"\(streamed)\" differs from the final reply")
        }
        if let tool, !(sawToolStart && sawToolComplete) {
            throw LiveScenarioError("The \(tool) tool did not start and complete")
        }
    }

    /// Every sequenced event arrives exactly once and in order across a reconnect.
    func expectContiguousSequence() throws {
        guard let first = sequence.first, sequence == Array(first..<(first + sequence.count)) else {
            throw LiveScenarioError("Session events were lost, duplicated or reordered: \(sequence)")
        }
    }
}

/// Calls the harness control endpoint that severs sockets or restarts the tagged server.
struct FaultControl: Sendable {
    let base: URL

    func drop(holdMilliseconds: Int) async throws {
        try await post("drop", query: [URLQueryItem(name: "hold_ms", value: String(holdMilliseconds))])
    }

    func blackhole() async throws {
        try await post("blackhole", query: [])
    }

    func restart() async throws {
        try await post("restart", query: [])
    }

    /// Sends the run's measured coverage evidence.
    func report(_ observations: [String: [String]]) async throws {
        try await post("report", query: [], body: JSONEncoder().encode(observations))
    }

    private func post(_ path: String, query: [URLQueryItem], body: Data? = nil) async throws {
        guard var components = URLComponents(url: base.appendingPathComponent(path), resolvingAgainstBaseURL: false)
        else { throw LiveScenarioError("Invalid control URL") }
        components.queryItems = query.isEmpty ? nil : query
        guard let url = components.url else { throw LiveScenarioError("Invalid control URL") }
        var request = URLRequest(url: url, timeoutInterval: 120)
        request.httpMethod = "POST"
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (_, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 204 else {
            throw LiveScenarioError("Harness control \(path) failed")
        }
    }
}

private actor Counter {
    private(set) var value = 0

    func increment() -> Int {
        value += 1
        return value
    }
}

private actor StateLog {
    private var states: [GatewayConnectionState] = []

    func isConnected(afterReconnects previous: Int) -> Bool {
        states.last == .connected && reconnects() > previous
    }

    func append(_ state: GatewayConnectionState) { states.append(state) }

    /// Completed reconnects: a reconnecting run followed by a connected state.
    func reconnects() -> Int {
        var count = 0
        var reconnecting = false
        for state in states {
            if case .reconnecting = state { reconnecting = true }
            if reconnecting && state == .connected {
                count += 1
                reconnecting = false
            }
        }
        return count
    }
}

private struct StaticTicketAuth: HermesAuth {
    let ticket: String

    func credential(baseURL: URL, http: any HTTPTransport) async throws -> GatewayCredential {
        .ticket(ticket, headers: [:])
    }
}

private func withDeadline<T: Sendable>(
    _ limit: Duration, _ waitingFor: String, _ operation: @escaping @Sendable () async throws -> T
) async throws -> T {
    try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await operation() }
        group.addTask {
            try await Task.sleep(for: limit)
            throw LiveScenarioError("Timed out waiting for \(waitingFor)")
        }
        defer { group.cancelAll() }
        guard let first = try await group.next() else { throw LiveScenarioError("No result for \(waitingFor)") }
        return first
    }
}

File: kotlin/src/live/kotlin/hermes/api/live/LiveScenarios.kt
package hermes.api.live

import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive
import hermes.api.generated.gateway.ApprovalChoice
import hermes.api.generated.gateway.ApprovalResult
import hermes.api.generated.gateway.ClarifyResult
import hermes.api.generated.gateway.GatewayEventPayload
import hermes.api.generated.gateway.MessageCompletePayloadText
import hermes.api.generated.gateway.PingParams
import hermes.api.generated.gateway.PromptSubmitParams
import hermes.api.generated.gateway.ServerRequest
import hermes.api.generated.gateway.ServerRequestResult
import hermes.api.generated.gateway.SessionCloseParams
import hermes.api.generated.gateway.SessionCreateParams
import hermes.api.generated.gateway.SessionListParams
import hermes.api.generated.rest.ProfilesSetActiveRequest
import hermes.api.runtime.GatewayConnection
import hermes.api.runtime.GatewayConnectionState
import hermes.api.runtime.GatewaySessionRecovery
import hermes.api.runtime.GatewayCredential
import hermes.api.runtime.GatewayHTTPTransport
import hermes.api.runtime.GatewayLogger
import hermes.api.runtime.GatewayNetworkMonitor
import hermes.api.runtime.GatewayTransport
import hermes.api.runtime.HermesAuth
import hermes.api.runtime.HermesGateway
import hermes.api.runtime.HermesGatewayConfiguration
import hermes.api.runtime.HermesGatewayException
import hermes.api.runtime.HermesREST
import hermes.api.runtime.HermesRESTConfiguration
import hermes.api.runtime.KtorGatewayTransport
import hermes.api.runtime.KtorRESTTransport
import hermes.api.runtime.LocalTokenAuth
import hermes.api.runtime.Patch

/** Live scenario inputs, read from `HERMES_LIVE_*` variables (or Android instrumentation arguments).
 *  Platforms pass their own network monitor and logger so the scenarios exercise them on the device. */
public class LiveScenarioEnvironment(
    values: Map<String, String>,
    public val networkMonitor: GatewayNetworkMonitor? = null,
    public val logger: GatewayLogger = GatewayLogger.None,
) {
    public val url: URI = URI(values["HERMES_LIVE_URL"] ?: throw LiveScenarioFailure("HERMES_LIVE_URL is required"))
    public val token: String? = values["HERMES_LIVE_TOKEN"]?.takeIf { it.isNotEmpty() }
    private val ticket: String? = values["HERMES_LIVE_TICKET"]?.takeIf { it.isNotEmpty() }
    public val lifecycle: Boolean = values["HERMES_LIVE_LIFECYCLE"] == "1"
    public val control: URI? = values["HERMES_LIVE_CONTROL"]?.takeIf { it.isNotEmpty() }?.let(::URI)

    public val auth: HermesAuth = when {
        ticket != null -> object : HermesAuth {
            override suspend fun credential(baseURI: URI, http: GatewayHTTPTransport): GatewayCredential =
                GatewayCredential.Ticket(ticket, emptyMap())
        }
        token != null -> LocalTokenAuth(token)
        else -> throw LiveScenarioFailure("HERMES_LIVE_TICKET or HERMES_LIVE_TOKEN is required")
    }
}

public class LiveScenarioFailure(message: String) : Exception(message)

/** Fixture replies produced by the harness model stub (`harness/stub_llm.py`). */
private object Fixture {
    const val REPLY = "HermesAPI fixture reply."
    const val CLARIFY_PROMPT = "Ask which release channel to use for HermesAPI."
    const val CLARIFY_REPLY = "HermesAPI stable release selected."
    const val APPROVAL_PROMPT = "Try the fixture cleanup command and report whether it was approved."
    const val APPROVAL_COMMAND = "rm -rf /tmp/hermes-api-fixture-approval-target"
    const val APPROVAL_REPLY = "HermesAPI approval denied as expected."
    const val RECONNECT_PROMPT = "Stream the HermesAPI reconnect fixture slowly."
    val RECONNECT_REPLY = (1..16).joinToString(" ") { "part" + it.toString().padStart(2, '0') }
}

public object LiveScenarios {
    /** Runs the smoke scenario and, when the harness exposes its control endpoint, the reconnect scenarios. */
    public suspend fun run(environment: LiveScenarioEnvironment) {
        val observations = LiveObservations()
        smoke(environment, observations)
        val control = environment.control ?: return
        val faults = FaultControl(control)
        if (environment.lifecycle) reconnect(environment, faults, observations)
        // Only a fully passing run reports what it exercised.
        faults.report(observations.report())
    }

    private suspend fun smoke(environment: LiveScenarioEnvironment, observations: LiveObservations) {
        val ktor = KtorGatewayTransport()
        val gateway = HermesGateway(HermesGatewayConfiguration(
            environment.url, environment.auth, ObservingTransport(ktor, observations), httpTransport = ktor,
            networkMonitor = environment.networkMonitor, logger = environment.logger,
        ))
        gateway.setServerRequestHandler { request ->
            when (request) {
                is ServerRequest.Clarify -> {
                    val questions = (request.params.questions as? Patch.Value)?.value
                        ?: throw LiveScenarioFailure("Missing clarification questions")
                    check(questions.size == 1 && questions[0].question == "Which release channel?") {
                        "Unexpected clarification question"
                    }
                    ServerRequestResult.Clarify(ClarifyResult(answers = mapOf(questions[0].qid to "Stable")))
                }
                is ServerRequest.Approval -> {
                    check(request.params.command == Fixture.APPROVAL_COMMAND) { "Unexpected approval command" }
                    ServerRequestResult.Approval(ApprovalResult(choice = ApprovalChoice.Deny))
                }
                else -> throw LiveScenarioFailure("Unexpected server request")
            }
        }
        try {
            gateway.connect()
            check(gateway.methods.ping(PingParams()).pong) { "Gateway ping returned false" }
            gateway.methods.gateway.capabilities(PingParams())
            if (environment.lifecycle) lifecycle(gateway, environment)
        } finally {
            gateway.disconnect()
            ktor.close()
        }
    }

    private suspend fun lifecycle(gateway: HermesGateway, environment: LiveScenarioEnvironment): Unit = coroutineScope {
        // Hermes broadcasts sessions.changed (at most every 2 s) once a turn writes the session store.
        val listChanged = async(start = CoroutineStart.UNDISPATCHED) {
            gateway.events.first { it.payload is GatewayEventPayload.SessionsChanged }
        }
        val session = gateway.methods.session.create(SessionCreateParams(
            cwd = Patch.Value("/tmp"), title = Patch.Value("HermesAPI live session"), closeOnDisconnect = true,
        ))
        runTurn(gateway, session.sessionId, "Reply with a short greeting.", null).expectReply(Fixture.REPLY)
        deadline(15_000, "a sessions.changed broadcast") { listChanged.await() }
        runTurn(gateway, session.sessionId, Fixture.CLARIFY_PROMPT, "clarify").expectReply(Fixture.CLARIFY_REPLY)
        val denied = runTurn(gateway, session.sessionId, Fixture.APPROVAL_PROMPT, "terminal")
        denied.expectReply(Fixture.APPROVAL_REPLY)
        val result = denied.toolResult as? JsonObject
        check(result?.get("status")?.jsonPrimitive?.content == "blocked" &&
            result["exit_code"]?.jsonPrimitive?.content == "-1") { "Denied terminal command was not blocked" }
        gateway.methods.session.list(SessionListParams())
        check(gateway.methods.session.close(SessionCloseParams(session.sessionId)).closed) {
            "Gateway session did not close"
        }
        val token = environment.token ?: throw LiveScenarioFailure("REST smoke needs the local token")
        val restTransport = KtorRESTTransport()
        try {
            val rest = HermesREST(HermesRESTConfiguration(environment.url,
                headers = { mapOf("X-Hermes-Session-Token" to token) }, transport = restTransport))
            val voice = rest.methods.audio.voiceLiveStatus(profile = "default")
            check(voice.ok && voice.mode == "chained" && voice.model.isNotEmpty() && voice.voice.isNotEmpty()) {
                "Unexpected voice status"
            }
            val profile = rest.methods.profiles.active()
            check(profile.active == "default" && profile.current == "default") { "Unexpected active profile" }
            val selected = rest.methods.profiles.setActive(ProfilesSetActiveRequest("default"))
            check(selected.ok && selected.active == "default") { "Could not select the default profile" }
            check(rest.methods.sessions.emptyCount(profile = "default").count >= 0) { "Invalid empty session count" }
        } finally {
            restTransport.close()
        }
    }

    /** Loses the socket mid-stream, silently, and with a question open, then resumes the session after a
     *  server restart. The heartbeat is shortened so a stalled socket is detected within seconds. */
    private suspend fun reconnect(
        environment: LiveScenarioEnvironment, faults: FaultControl, observations: LiveObservations,
    ): Unit = coroutineScope {
        val ktor = KtorGatewayTransport()
        val transport = ObservingTransport(ktor, observations)
        val gateway = HermesGateway(HermesGatewayConfiguration(
            environment.url, environment.auth, transport, httpTransport = ktor, reconnectDelayMillis = { 250 },
            heartbeatIntervalMillis = 1_000, heartbeatDeadlineMillis = 4_000,
            networkMonitor = environment.networkMonitor, logger = environment.logger,
        ))
        suspend fun awaitReconnect(after: Int) {
            while (transport.reconnects <= after || gateway.connectionStates.value != GatewayConnectionState.Connected) {
                delay(100)
            }
        }
        val clarifications = AtomicInteger()
        gateway.setServerRequestHandler { request ->
            val questions = ((request as? ServerRequest.Clarify)?.params?.questions as? Patch.Value)?.value
                ?: throw LiveScenarioFailure("Unexpected server request during reconnect")
            if (clarifications.incrementAndGet() == 1) {
                // Drop the socket while the question is open; the rebind must re-deliver it.
                faults.drop(holdMillis = 1_000)
                delay(60_000)
            }
            ServerRequestResult.Clarify(ClarifyResult(answers = mapOf(questions[0].qid to "Stable")))
        }
        try {
            gateway.connect()
            val session = gateway.methods.session.create(SessionCreateParams(
                cwd = Patch.Value("/tmp"), title = Patch.Value("HermesAPI reconnect session"), closeOnDisconnect = false,
            ))

            val streamed = runTurn(gateway, session.sessionId, Fixture.RECONNECT_PROMPT, null) {
                faults.drop(holdMillis = 1_500)
            }
            streamed.expectReply(Fixture.RECONNECT_REPLY)
            if (streamed.replayedEvents == 0) throw LiveScenarioFailure("No event reached the client through reconnect replay")
            streamed.expectContiguousSequence()
            if (transport.reconnects < 1) throw LiveScenarioFailure("Gateway never reconnected")

            // A stalled socket reports nothing; the default heartbeat must replace it before the
            // server reaps the session, which the next turn on the same session then proves.
            val reconnectsBeforeStall = transport.reconnects
            faults.blackhole()
            deadline(20_000, "the heartbeat to replace a silently stalled socket") { awaitReconnect(reconnectsBeforeStall) }

            val clarified = runTurn(gateway, session.sessionId, Fixture.CLARIFY_PROMPT, "clarify")
            clarified.expectReply(Fixture.CLARIFY_REPLY)
            clarified.expectContiguousSequence()
            if (clarifications.get() != 2) {
                throw LiveScenarioFailure("The open clarification was not re-delivered after reconnect")
            }

            // A restart loses every live session; the gateway resumes this one from Hermes' storage.
            val reconnectsBeforeRestart = transport.reconnects
            val recovery = async(start = CoroutineStart.UNDISPATCHED) { gateway.sessionRecoveries.first() }
            faults.restart()
            val resumed = deadline(60_000, "the restarted server's session recovery") { recovery.await() }
            if (resumed !is GatewaySessionRecovery.Resumed || resumed.previousSessionId != session.sessionId ||
                resumed.storedSessionId != session.storedSessionId) {
                throw LiveScenarioFailure("Expected ${session.sessionId} to be resumed, got $resumed")
            }
            deadline(60_000, "the gateway to reconnect after the restart") { awaitReconnect(reconnectsBeforeRestart) }
            runTurn(gateway, resumed.sessionId, "Reply with a short greeting.", null).expectReply(Fixture.REPLY)
        } finally {
            gateway.disconnect()
            ktor.close()
        }
    }

    private suspend fun runTurn(
        gateway: HermesGateway, sessionId: String, prompt: String, tool: String?,
        onFirstDelta: (suspend () -> Unit)? = null,
    ): Turn = coroutineScope {
        val turn = Turn(tool)
        val completion = async(start = CoroutineStart.UNDISPATCHED) {
            // Generous: Hermes builds the agent on the first turn, which is slow on a cold CI runner.
            deadline(120_000, "the turn for \"$prompt\" to complete") {
                gateway.events.first { event ->
                    if (event.sessionId != sessionId) return@first false
                    event.seq?.let(turn.sequence::add)
                    if (event.replayed) turn.replayedEvents++
                    when (val payload = event.payload) {
                        GatewayEventPayload.MessageStart -> turn.sawStart = true
                        is GatewayEventPayload.ToolStart -> if (payload.payload.name == tool) turn.sawToolStart = true
                        is GatewayEventPayload.ToolComplete -> if (payload.payload.name == tool) {
                            turn.sawToolComplete = true
                            turn.toolResult = payload.payload.result
                        }
                        is GatewayEventPayload.MessageDelta -> {
                            val first = turn.streamed.isEmpty()
                            turn.streamed.append(payload.payload.text)
                            if (first) onFirstDelta?.invoke()
                        }
                        is GatewayEventPayload.MessageComplete ->
                            turn.reply = (payload.payload.text as? MessageCompletePayloadText.StringValue)?.value
                        is GatewayEventPayload.Error -> throw LiveScenarioFailure("Gateway reported an error during \"$prompt\"")
                        else -> Unit
                    }
                    event.type == "message.complete"
                }
            }
        }
        val submitted = gateway.methods.prompt.submit(PromptSubmitParams(sessionId = sessionId, text = JsonPrimitive(prompt)))
        if (submitted.status == null) throw LiveScenarioFailure("Gateway rejected the prompt")
        completion.await()
        turn
    }
}

private class Turn(val tool: String?) {
    var sawStart = false
    var sawToolStart = false
    var sawToolComplete = false
    var toolResult: JsonElement? = null
    val streamed = StringBuilder()
    var reply: String? = null
    val sequence = mutableListOf<Long>()
    var replayedEvents = 0

    fun expectReply(expected: String) {
        if (reply != expected) throw LiveScenarioFailure("Expected reply \"$expected\", got \"${reply ?: "nothing"}\"")
        if (!sawStart) throw LiveScenarioFailure("No message.start before \"$expected\"")
        if (streamed.toString().trim() != reply) {
            throw LiveScenarioFailure("Streamed text \"$streamed\" differs from the final reply")
        }
        if (tool != null && !(sawToolStart && sawToolComplete)) {
            throw LiveScenarioFailure("The $tool tool did not start and complete")
        }
    }

    /** Every sequenced event arrives exactly once and in order across a reconnect. */
    fun expectContiguousSequence() {
        val first = sequence.firstOrNull()
        if (first == null || sequence != (first until first + sequence.size).toList()) {
            throw LiveScenarioFailure("Session events were lost, duplicated or reordered: $sequence")
        }
    }
}

/** Calls the harness control endpoint that severs sockets or restarts the tagged server. */
private class FaultControl(private val base: URI) {
    suspend fun drop(holdMillis: Int) = post("drop?hold_ms=$holdMillis")

    suspend fun blackhole() = post("blackhole")

    suspend fun restart() = post("restart")

    /** Sends the run's measured coverage evidence. */
    suspend fun report(json: String) = post("report", json)

    private suspend fun post(path: String, json: String? = null): Unit = withContext(Dispatchers.IO) {
        val connection = URL(base.resolve("/$path").toString()).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 10_000
            connection.readTimeout = 120_000
            if (json != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(json.toByteArray()) }
            }
            if (connection.responseCode != 204) throw LiveScenarioFailure("Harness control $path failed")
        } finally {
            connection.disconnect()
        }
    }
}

private suspend fun <T> deadline(millis: Long, waitingFor: String, block: suspend () -> T): T = try {
    withTimeout(millis) { block() }
} catch (_: kotlinx.coroutines.TimeoutCancellationException) {
    throw LiveScenarioFailure("Timed out waiting for $waitingFor")
}
